File: verification/offline_exf_seaice/input/gendata.m

kwr=1; kprt=0;
nx=80; ny=42; nr=3; nt=1;

xc=[1:nx]; xc=xc-mean(xc);
yc=[1:ny]-.5;

%------------------------------------------------------

windx=10.;
H0=-100.;

namf='channel.bin';
depth=H0*ones(nx,ny); depth(:,1)=0.;
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,depth,'real*8'); fclose(fid);
end

namf='bathy_3c.bin';
msk=abs(xc)'*ones(1,ny)+ones(nx,1)*yc;
depth=H0*ones(nx,ny); depth(:,1)=0.;
depth(find(msk < 24))=0.;
y2d=ones(nx,1)*yc;
depth(find(y2d > ny/2))=H0;
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,depth,'real*8'); fclose(fid);
end

namf='windx.bin';
wnd=windx*ones(nx,ny,nt);
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,wnd,'real*8'); fclose(fid);
end

%- file name convention: "const_{xx}.bin" <-> uniform value = xx (in percent)
namf='const_00.bin';
fld=0*ones(nx,ny,nt);
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,fld,'real*8'); fclose(fid);
end

namf='const100.bin'; w0=1.;
var=w0*ones(nx,ny);
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,var,'real*8'); fclose(fid);
end

namf='const+20.bin'; w0=0.2;
var=w0*ones(nx,ny);
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,var,'real*8'); fclose(fid);
end

%namf='const+40.bin';
%u0=0.4;
%var=u0*ones(nx,ny);
%if kwr > 0,
% fprintf('write to file: %s\n',namf);
% fid=fopen(namf,'w','b'); fwrite(fid,var,'real*8'); fclose(fid);
%end

%namf='const-10.bin';
%v0=-0.1;
%var=v0*ones(nx,ny);
%if kwr > 0,
% fprintf('write to file: %s\n',namf);
% fid=fopen(namf,'w','b'); fwrite(fid,var,'real*8'); fclose(fid);
%end
%------------------------------------------------------

namf='ice0_area.bin'; iceC0=1.;
iceConc=iceC0*ones(nx,ny); iceConc(:,1)=0;
iceConc(:,2)=0.00*iceC0;
iceConc(:,3)=0.10*iceC0;
iceConc(:,end)  =0.00*iceC0;
iceConc(:,end-1)=0.01*iceC0;
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,iceConc,'real*8'); fclose(fid);
end

namf='ice0_heff.bin'; iceH0=0.2;
iceVol=iceConc*iceH0;
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,iceVol,'real*8'); fclose(fid);
end

%------------------------------------------------------

dsw0=100;
namf=['dsw_',int2str(dsw0),'.bin'];
fld=dsw0*ones(nx,ny,nt);
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,fld,'real*8'); fclose(fid);
end

dlw0=250;
namf=['dlw_',int2str(dlw0),'.bin'];
fld=dlw0*ones(nx,ny,nt);
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,fld,'real*8'); fclose(fid);
end

cel2K=273.15; dtx=4; %- dtx = amplitude of air temp variations in X-dir
ta_x=cel2K + dtx*sin(pi*(1+2*xc'/nx));
ta=repmat(ta_x,[1 ny nt]);
namf=['tair_',int2str(dtx),'x.bin'];
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,ta,'real*8'); fclose(fid);
end;

cvapor_fac     =   640380.000  ;
cvapor_exp     =     5107.400  ;
atmrho         =        1.200  ;
rh=70; %- specific humid <--> 70.% relative humid
tmpbulk = cvapor_fac*exp(-cvapor_exp./ta_x);
qa_x = (rh/100.)*tmpbulk/atmrho ;
qa=repmat(qa_x,[1 ny nt]);
namf=['qa',int2str(rh),'_',int2str(dtx),'x.bin'];
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,qa,'real*8'); fclose(fid);
end;

%- salinity
sCst=30;
so=sCst*ones(nx,ny,nt);
namf='socn.bin';
%if kwr > 0,
% fprintf('write to file: %s\n',namf);
% fid=fopen(namf,'w','b'); fwrite(fid,so,'real*8'); fclose(fid);
%end;

muTf = 5.4e-2;
tfreeze=-muTf*sCst;
fprintf('T-freeze = %10.6f\n',tfreeze);
%- parabolic profile in Y, max @ j=4, min @ j=ny, amplitude=1.K
to_y=(yc-3.5)/(ny-4);
to_y=tfreeze+0.5-to_y.*to_y;
  mnV=min(to_y); MxV=max(to_y); Avr=mean(to_y(2:end));
  fprintf(' SST* av,mn,Mx: %9.6f , %9.6f , %9.6f , %9.6f\n',Avr,mnV,MxV,MxV-mnV);
to=repmat(to_y,[nx 1 nt]);
namf='tocn.bin';
if kwr > 0,
 fprintf('write to file: %s\n',namf);
 fid=fopen(namf,'w','b'); fwrite(fid,to,'real*8'); fclose(fid);
end;

%-- make some plots to check: ----------------

hScal=[-1.1 0.1]*abs(H0);
figure(1); clf;
subplot(211);
var=depth;
imagesc(xc,yc,var'); set(gca,'YDir','normal');
%caxis(hScal);
%change_colmap(-1);
colorbar;
title('Depth [m]');

subplot(413);
var=depth;
j1=2;
j2=ny/2;
j3=j2+1;
plot(xc,var(:,j1),'k-')
hold on; j=j+1;
plot(xc,var(:,j2),'ro-')
plot(xc,var(:,j3),'b-')
hold off;
axis([-nx/2 nx/2 hScal]);
grid
legend(int2str(j1),int2str(j2),int2str(j3));
title('Depth @ j= cst');

subplot(414);
i=nx/2;
plot(yc,var(i,:),'k-')
axis([0 ny H0*1.1 -H0*.1]);
grid
title(['Depth @ i=',int2str(i)]);

%--
dewPt=(qa_x*atmrho)/cvapor_fac;
dewPt=-cvapor_exp./log(dewPt);

figure(2);clf;
subplot(311)
P(1)=plot(xc,ta_x-cel2K,'r-'); hold on;
P(2)=plot(xc,dewPt-cel2K,'b-');
P(3)=plot(xc,tfreeze*ones(nx,1),'k-');
set(P,'LineWidth',1);
hold off; AA=axis;
axis([-nx/2 nx/2 AA(3:4)]);
legend('ta','dew');
grid
xlabel('X')
title(['Air Temp (^oC): del-Temp-X = ',int2str(dtx),' , RH= ',int2str(rh)]);
subplot(312)
P(1)=plot(yc,to_y,'b-'); hold on;
P(2)=plot(yc,tfreeze*ones(ny,1),'k-');
set(P,'LineWidth',1);
hold off; AA=axis;
L=line([1 1],AA(3:4)); set(L,'LineWidth',2.,'Color',[0 0 0]);
axis([0 ny AA(3:4)]);
grid
xlabel('Y')
title('Ocean Temp ^oC');

subplot(313)
var=iceConc(1,:);
P(1)=semilogy(yc,var,'b-x'); hold on;
%plot(yc,var,'b-x'); hold on;
var=iceVol(1,:);
P(2)=semilogy(yc,var,'r-x');
%plot(yc,var,'r-x');
set(P,'LineWidth',1);
hold off; AA=axis;
L=line([1 1],AA(3:4)); set(L,'LineWidth',2.,'Color',[0 0 0]);
axis([0 ny [0 2]*iceC0]);
grid
xlabel('Y')
legend('iceC','hEff','Location','South');
title('Initial ice in Channel : y-section');
%-----
if kprt == 1, f=2;
 namfig=sprintf('forcing_%2.2i',f);
 fprintf([' print fig= %2i to file: ',namfig,' '],f);
 set(f,'PaperOrientation','portrait')
%set(f,'PaperPosition',[0.25 1.5 6. 8.]);
 set(f,'PaperPosition',[0.25 1.5 5.25 7.]);
 print(f,'-depsc2',namfig); fprintf('\n');
end
%-----

figure(3);clf;
subplot(311)
var=iceConc; ccB=[-1 12]/10;
imagesc(xc,yc,var'); set(gca,'YDir','normal');
caxis(ccB);
%change_colmap(-1);
colorbar;
title('Ice Concentration in Channel');

subplot(312)
var=iceVol; ccB=[-1 12]/50;
imagesc(xc,yc,var'); set(gca,'YDir','normal');
caxis(ccB);
%change_colmap(-1);
colorbar;
title('Effective ice thickness in Channel');

subplot(313)
var=iceConc(1,:);
%plot(yc,var,'b-x'); hold on;
semilogy(yc,var,'b-x'); hold on;
var=iceVol(1,:);
%plot(yc,var,'r-x'); hold off;
semilogy(yc,var,'r-x'); hold on;
AA=axis; axis([0 ny [0 2]*iceC0]);
grid
legend('iceC','hEff','Location','South');
title('Initial ice in Channel : y-section');
%-----

return

File: verification/seaice_itd/input/gendata.py
# simple python script to generate input data
# this script is very similar to $ROOTDIR/verification/offline_exf_seaice/input/gendata.m
#
import numpy as np

kwr, kprt =1, 0
nx, ny, nr, nt = 80, 42, 3, 1

xc = np.arange(nx)
xc = xc - xc.mean()
yc = np.arange(ny)+0.5

# ------------------------------------------------------

def writefield(fname,data):
    import sys
    print 'write to file: '+fname
    if sys.byteorder == 'little': data.byteswap(True)
    fid = open(fname,"wb")
    data.tofile(fid)
    fid.close()


windx =   10.
H0    = -100. 

# pure channel
namf='channel.bin'
depth=H0*np.ones((ny,nx),dtype='float64'); depth[0,:] = 0.
#if kwr > 0: writefield(namf,depth)

# triangular obstacle
namf='bathy_3c.bin';
depth=H0*np.ones((ny,nx),dtype='float64'); depth[0,:]=0.;
msk=np.tile(np.abs(xc),(ny,1)) + np.tile(yc,(nx,1)).T
depth[msk < 24]=0.;
y2d=np.tile(yc,(nx,1)).T
depth[y2d > ny/2.]=H0;
if kwr > 0: writefield(namf,depth)

# wind field
namf='windx.bin';
wnd=windx*np.ones((nt,ny,nx),dtype='float64')
if kwr > 0: writefield(namf,wnd)

#- file name convention: "const_{xx}.bin" <-> uniform value = xx (in percent)
namf='const_00.bin';
fld=np.zeros((nt,ny,nx),dtype='float64')
if kwr > 0: writefield(namf,fld)

namf='const100.bin'; w0=1.;
var=w0*np.ones((nt,ny,nx),dtype='float64')
if kwr > 0: writefield(namf,var)

namf='const+20.bin'; w0=0.2;
var=w0*np.ones((nt,ny,nx),dtype='float64')
#if kwr > 0: writefield(namf,var)

namf='heff_quartic.bin'
hf_y = 8.*(yc/ny)**4
hf_y.resize((1,ny,1))
hf=np.tile(hf_y,(nt,1,nx))
if kwr > 0: writefield(namf,hf)

#------------------------------------------------------

namf='ice0_area.bin'; iceC0=1.;
iceConc=iceC0*np.ones((ny,nx),dtype='float64'); iceConc[0,:]=0;
iceConc[1,:] =0.00*iceC0;
iceConc[2,:] =0.10*iceC0;
iceConc[-1,:]=0.00*iceC0;
iceConc[-2,:]=0.01*iceC0;
#if kwr > 0: writefield(namf,iceConc)

namf='ice0_heff.bin'; iceH0=0.2;
iceVol=iceConc*iceH0;
#if kwr > 0: writefield(namf,iceVol)

#------------------------------------------------------

dsw0=100;
namf='dsw_'+str(dsw0)+'.bin'
fld=dsw0*np.ones((nt,ny,nx),dtype='float64')
if kwr > 0: writefield(namf,fld)

dlw0=250;
namf='dlw_'+str(dlw0)+'.bin'
fld=dlw0*np.ones((nt,ny,nx),dtype='float64')
if kwr > 0: writefield(namf,fld)

cel2K=273.15; dtx=4; #- dtx = amplitude of air temp variations in X-dir
ta_x=cel2K + dtx*np.sin(np.pi*(1+2*xc/nx));
ta=np.tile(ta_x,(nt,ny,1))
namf='tair_'+str(dtx)+'x.bin'
if kwr > 0: writefield(namf,ta)

cvapor_fac     =   640380.000
cvapor_exp     =     5107.400
atmrho         =        1.200
rh=70; #- specific humid <--> 70.% relative humid
tmpbulk = cvapor_fac*np.exp(-cvapor_exp/ta_x);
qa_x = (rh/100.)*tmpbulk/atmrho
qa=np.tile(qa_x,(nt,ny,1))
namf='qa'+str(rh)+'_'+str(dtx)+'x.bin'
if kwr > 0: writefield(namf,qa)

#- salinity
sCst=30
so=sCst*np.ones((nt,ny,nx),dtype='float64')
namf='socn.bin'
#if kwr > 0: writefield(namf,so)

muTf = 5.4e-2;
tfreeze=-muTf*sCst;
print 'T-freeze = {0:10.6f}'.format(tfreeze)
#- parabolic profile in Y, max @ j=4, min @ j=ny, amplitude=1.K
to_y=(yc-3.5)/(ny-4)
to_y=tfreeze+0.5-to_y*to_y
mnV=to_y.min(); MxV=to_y.max(); Avr=to_y[1:].mean()
print ' SST* av,mn,Mx: {0:9.6f} , {1:9.6f} , {2:9.6f} , {3:9.6f}'.format(Avr,mnV,MxV,MxV-mnV)
to_y.resize((1,ny,1))
to=np.tile(to_y,(nt,1,nx))
namf='tocn.bin';
if kwr > 0: writefield(namf,to)

#-- make some plots to check: ----------------

import matplotlib.pyplot as plt
hScal=np.asarray([-1.1, 0.1])*np.abs(H0);
xg = xc-.5
yg = yc-.5
plt.figure(1); plt.clf()
plt.subplot(211)
var=depth
plt.pcolormesh(xg,yg,var)
plt.colorbar()
plt.title('Depth [m]')

plt.subplot(413)
var=depth
j1=1
j2=ny/2-1
j3=j2+1
plt.plot(xc,var[j1,:],'k-',label=str(j1))
plt.plot(xc,var[j2,:],'ro-',label=str(j2))
plt.plot(xc,var[j3,:],'b-',label=str(j3))
plt.axis([-nx/2, nx/2, hScal[0], hScal[1]])
plt.grid()
plt.legend()
plt.title('Depth @ j= cst');

plt.subplot(414);
i=nx/2-1
plt.plot(yc,var[:,i],'k-')
plt.axis([0,ny,H0*1.1,-H0*.1]);
plt.grid()
plt.title(['Depth @ i=',str(i)]);

#--
dewPt=(qa_x*atmrho)/cvapor_fac;
dewPt=-cvapor_exp/np.log(dewPt);

plt.figure(2);plt.clf();
plt.subplot(311)
plt.plot(xc,ta_x-cel2K,'r-',label='ta',linewidth = 1)
plt.plot(xc,dewPt-cel2K,'b-',label='dew',linewidth = 1)
plt.plot(xc,tfreeze*np.ones((nx,1)),'k-',linewidth = 1)
AA=plt.axis()
plt.axis([-nx/2, nx/2, AA[2], AA[3]]);
plt.legend()
plt.grid()
plt.xlabel('X')
plt.title('Air Temp ($^\circ$C): del-Temp-X = '+str(dtx)+' , RH= '+str(rh));
plt.subplot(312)
plt.plot(yc,to_y.flatten(),'b-',linewidth = 1)
plt.plot(yc,tfreeze*np.ones((ny,1)),'k-',linewidth = 1)
AA=plt.axis()
plt.plot([1,1],AA[2:],'k',linewidth=2.)
plt.axis([0, ny, AA[2], AA[3]])
plt.grid()
plt.xlabel('Y')
plt.title('Ocean Temp $^\circ$C');

plt.subplot(313)
var=iceConc[:,0]
plt.semilogy(yc,var,'b-x',linewidth = 1,label='iceC')
var=iceVol[:,0]
plt.semilogy(yc,var,'r-x',linewidth = 1,label='hEff')
AA=plt.axis()
plt.plot([1,1],AA[2:],'k',linewidth=2.)
plt.axis([0, ny, 0, 2*iceC0])
plt.grid()
plt.xlabel('Y')
plt.legend(loc='lower center')
plt.title('Initial ice in Channel : y-section');
#
if kprt == 1:
    f=2
    namfig='forcing_{0:02g}'.format(f)
    print ' print fig= {0:2g} to file: '.format(f) + namfig
    print(f,'-depsc2',namfig); fprintf('\n');

    
plt.figure(3);plt.clf();
plt.subplot(311)
plt.pcolormesh(xg,yg,iceConc,vmin=-0.1, vmax=1.2)
plt.colorbar()
plt.title('Ice Concentration in Channel');

plt.subplot(312)
plt.pcolormesh(xg,yg,iceVol,vmin=-1./50., vmax=12./50.)
plt.colorbar()
plt.title('Effective ice thickness in Channel');

plt.subplot(313)
plt.semilogy(yc,iceConc[:,0],'b-x',label='iceC')
plt.semilogy(yc,iceVol[:,0],'r-x',label='hEff')
AA=plt.axis(); plt.axis([0,ny,0,2*iceC0]);
plt.grid()
plt.legend(loc='lower center')
plt.title('Initial ice in Channel : y-section');

plt.show()
